File: php/199-binary-tree-right-side-view.php
<?php
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($value) { $this->val = $value; }
 * }
 */
class Solution 
{
    /**
     * @param TreeNode $root
     * @return Integer[]
     */
    public function rightSideView($root) 
    {
        $queue[] = [$root, 0];
        $data = [];
        while ($queue) {
            [$node, $dir, $level] = array_shift($queue);
            if (!isset($data[$level])) {
                $data[$level] = $node->val;
            }
            if ($node->right) {
                $queue[] = [$node->right, 1, $level+1];
            }
            if ($node->left) {
                $queue[] = [$node->left, 0, $level+1];
            }
        }
        return $data;
    }
}

File: go/199-binary-tree-right-side-view.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type QueueNode struct {
	node  *TreeNode
	level int
}

func rightSideView(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}
	queue := []QueueNode{}
	queue = append(queue, QueueNode{root, 0})
	resMap := make(map[int]int)
	for len(queue) > 0 {
		q := queue[0]
		queue = queue[1:]
		if _, ok := resMap[q.level]; !ok {
			resMap[q.level] = q.node.Val
		}
		if q.node.Right != nil {
			queue = append(queue, QueueNode{q.node.Right, q.level + 1})
		}
		if q.node.Left != nil {
			queue = append(queue, QueueNode{q.node.Left, q.level + 1})
		}
	}
	var res []int
	for i := 0; i < len(resMap); i++ {
		res = append(res, resMap[i])
	}
	return res
}
